t/lib/pause_1999/Test/SiteModel/Parser.pm

```diff
@@ -0,0 +1,109 @@
+package pause_1999::Test::SiteModel::Parser;
+
+use Moose::Role;
+use HTML::TreeBuilder;
+use YAML::XS qw/Load/;
+requires 'mech';
+
+# See the SiteModel for the basic justification
+#
+# Extend the SiteModel to be able to ->parse() certain pages
+#
+# Add the URL fragment to %pages, with a list of parsers to be used to extract
+# data from the page. Parsers are by name in %parsers, and receive a
+# HTML::TreeBuilder object.
+
+our %pages = (
+    homepage              => ['basic'],
+    delete_files          => [qw/basic author_directory file_list/],
+    email_for_admin       => [qw/basic email_for_admin/],
+    email_for_admin__yaml => [qw/yaml/],
+    show_files            => [qw/basic author_directory file_list/],
+);
+
+our %parsers = (
+    author_directory => sub {
+        my $tree = shift;
+        my ($directory) = ( $tree->as_text =~ m!Files in directory (.+?) ! );
+        return $directory;
+    },
+    basic => sub {
+        my $tree = shift;
+        my $status_box = $tree->find_by_attribute( class => 'statusunencr' )
+            || $tree->find_by_attribute( class => 'statusencr' );
+
+        my ( $username, $email ) = ( $status_box->as_text =~ m/(.+) <(.+)>/ );
+
+        return {
+            username => $username,
+            email    => $email,
+        };
+    },
+    email_for_admin => sub {
+        my $tree           = shift;
+        my @all_tables     = $tree->look_down( _tag => 'table' );
+        my @content_tables = $all_tables[2]->look_down( _tag => 'table' );
+        my $author_table   = $content_tables[2];
+        my %authors
+            = map { length $_->as_text == 0 ? undef : $_->as_text }
+            $author_table->look_down( _tag => 'td' );
+        return \%authors;
+    },
+    file_list => sub {
+        my $tree = shift;
+        my $pre = $tree->look_down( _tag => 'pre' );
+        return [] unless $pre;
+        my @files = map {
+            my $line = $_;
+            if ( $line =~ m/([^ ]+)\s+(\d+)\s+([^<]+)/ ) {
+                { filename => $1, size => $2, date => $3 };
+            }
+            else {
+                ();
+            }
+        } split( m!<br />|\n!, $pre->as_HTML );
+        return \@files;
+    },
+    yaml => sub {
+        my ( $tree, $content ) = @_;
+        return Load $content;
```

Comment:
In list context, `Load` returns a list.  Although we can take it as *really* unlikely that we ever have more than one document in `$content`, but still, we should put this Load into scalar context with `scalar`, lest we screw up a hash horribly.